File: py/graph/strongly_connected_component/Gabow.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class Gabow:
    def search(self, graphs: List[VertexNode]) -> List[List[int]]:
        parent, visited, path, root = [-1 for _ in range(len(graphs))], [False for _ in range(len(graphs))], [], []
        result, aux = [], []
        for i in range(len(graphs)):
            if not ~parent[i]:
                self.__search(graphs, i, parent, visited, path, root, result, aux, 0)
        return result

    def __search(self, graphs: List[VertexNode], start: int, parent: List[int], visited: List[bool], path: List[int], root: List[int], result: List[List[int]], aux: List[int], count: int) -> int:
        parent[start] = count
        path.append(start)
        root.append(start)

        edge = graphs[start].next
        while edge:
            edgeIndex = edge.index
            if not ~parent[edgeIndex]:
                count = self.__search(graphs, edgeIndex, parent, visited, path, root, result, aux, count + 1)
            elif not visited[edgeIndex]:
                while parent[root[-1]] > parent[edgeIndex]:
                    root.pop()
            edge = edge.next

        if start != root[-1]:
            return count
        root.pop()

        index = path.pop()
        while index != start:
            aux.append(index)
            visited[index], index = True, path.pop()
        aux.append(start)
        visited[start] = True
        result.append(list(aux))
        aux.clear()
        return count

File: java/graph/strongly_connected_component/Gabow.java
package graph.strongly_connected_component;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Stack;

public class Gabow {
    public int[][] search(VertexNode[] graphs) {
        int[] parent = new int[graphs.length];
        Arrays.fill(parent, -1);
        boolean[] visited = new boolean[graphs.length];
        Arrays.fill(visited, false);
        Stack<Integer> path = new Stack<>();
        Stack<Integer> root = new Stack<>();

        List<int[]> result = new ArrayList<>();
        List<Integer> aux = new ArrayList<>();
        for (int i = 0; i < graphs.length; ++i) {
            if (0 == ~parent[i]) {
                search(graphs, i, parent, visited, path, root, result, aux, 0);
            }
        }
        return result.toArray(new int[0][]);
    }

    private int search(VertexNode[] graphs, int start, int[] parent, boolean[] visited, Stack<Integer> path, Stack<Integer> root, List<int[]> result, List<Integer> aux, int count) {
        parent[start] = count;
        path.push(start);
        root.push(start);

        EdgeNode edge = graphs[start].next;
        while (null != edge) {
            int edgeIndex = edge.index;
            if (0 == ~parent[edgeIndex]) {
                count = search(graphs, edgeIndex, parent, visited, path, root, result, aux, count + 1);
            } else if (!visited[edgeIndex]) {
                while (parent[root.peek()] > parent[edgeIndex]) {
                    root.pop();
                }
            }
            edge = edge.next;
        }

        if (start != root.peek()) {
            return count;
        }
        root.pop();

        int index = path.pop();
        while (index != start) {
            aux.add(index);
            visited[index] = true;
            index = path.pop();
        }
        aux.add(start);
        visited[start] = true;
        result.add(aux.stream().mapToInt(Integer::valueOf).toArray());
        aux.clear();
        return count;
    }

    public class EdgeNode {
        public int index = 0;
        public int weight = 0;
        public EdgeNode next = null;

        public EdgeNode(int index, int weight) {
            this.index = index;
            this.weight = weight;
        }
    }

    public class VertexNode {
        public int value = 0;
        public EdgeNode next = null;

        public VertexNode(int value) {
            this.value = value;
        }
    }
}
